### PR TITLE
Improve SQL operator signature errors (#17301)

### LINKED ISSUE #17301: Find a universal and cheap way to customize operator signature error messages
Default Calcite template for operator signature mismatches might be misleading: `Cannot apply 'NOT' to arguments of type 'NOT<DOUBLE>'. Supported form(s): 'NOT<BOOLEAN>'`. It refers to the signature/overload/form as a type and that might be misleading. Also, it's sometimes desirable to remove/add/customize some signatures in the supported forms list. Relevant comments:

https://github.com/hazelcast/hazelcast/pull/17183#pullrequestreview-444659460
https://github.com/hazelcast/hazelcast/pull/17183#pullrequestreview-446902967
https://github.com/hazelcast/hazelcast/pull/17183#pullrequestreview-446903114
https://github.com/hazelcast/hazelcast/pull/17183#pullrequestreview-454199466

### PR DESCRIPTION
**Pull request by devozerov**
This PR introduces the common handling of SQL operator signature errors. 

Before that the error message was in the form: `Cannot apply {0} to arguments of type {1}. Supported form(s): {2}`, where {0} was the operator name, {1} - signature, {2} - supported signatures. There were two problems with this error message:
1. Signature was used in a place where types must be used instead. This led to counterintuitive error messages, like `Cannot apply 'NOT' to arguments of type 'NOT<TINYINT>' ...`. What is `NOT<TINYINT>`? We do not have such type, we have `TINYINT`
1. Supported forms were constructed based on Calcite code, which is not necessarily on par with current Hazelcast SQL behavior and capabilities. For example, it may print `TIMESTAMP` as one of the supported forms for a comparison operator, while we do not support it in reality. We do not want to confuse our users.

This PR changes the error message to the following forms: 
1. Functions: `No function matches {0} name and argument types (you might need to add an explicit CAST)`
1. Operators: `No operator matches {0} name and argument types (you might need to add an explicit CAST)`
Where {0} is the operator's signature, e.g. `ABS(<INTEGER>)`, `INTEGER + STRING`, etc.

This approach is borrowed from Postgres. We do not have "supported form(s)" now, so no confusion. We also remove this unclear `... type 'NOT<TINYINT>' ...` part of the message.

To do this we hook into the validation process and inject our own implementation of `SqlCallBinding`, which throws custom error messages in the desired form. See `SqlCallBindingOverride` class.

Also one unrelated change to `CompositeIndexVisitor` to let FindBugs work.

Closes #17301

### CODE REVIEW

**Review by viliam-durina — APPROVED** [review]
I'm approving since I only recommended changing the local variable names and the error text.

File: hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlOperatorTable.java [hunk 1]
```diff
@@ -428,4 +449,60 @@ private HazelcastSqlOperatorTable() {
...
+
+                if (resolvedOperators.isEmpty()) {
+                    throw functionDoesNotExist(call);
+                }
+
+                assert resolvedOperators.size() == 1;
+
+                basicCall.setOperator(resolvedOperators.get(0));
+            } else if (call instanceof SqlCase) {
+                // TODO: Support CASE
```
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlOperatorTable.java:499]: No, we will remove this `TODO` soon, when `CASE` support is added (https://github.com/hazelcast/hazelcast/issues/17329)

File: hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastResources.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Error messages for parsing and validation stages.
+ */
+public interface HazelcastResources {
+
+    HazelcastResources RESOURCES = Resources.create(HazelcastResources.class);
+
+    @Resources.BaseMessage("Function ''{0}'' does not exist")
+    Resources.ExInst<CalciteException> functionDoesNotExist(String a0);
```
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastResources.java:31]: Fixed

File: hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastResources.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Resources.BaseMessage("No function matches {0} name and argument types (you might need to add an explicit CAST)")
+    Resources.ExInst<SqlValidatorException> canNotApplyOperandsToFunction(String a0);
+
+    @Resources.BaseMessage("No operator matches {0} name and argument types (you might need to add an explicit CAST)")
+    Resources.ExInst<SqlValidatorException> canNotApplyOperandsToOperator(String a0);
```
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastResources.java:37]: Will apply the templates as per our internal discussion:
`Cannot apply {1} to the {0} operator (consider adding an explicit CAST)`
`Cannot apply {1} to the {0} function (consider adding an explicit CAST)`

**Comment by devozerov** [comment]
Since we are already advanced our new expressions implementation a lot, this separate PR doesn't make much sense. Closing.

**Pull request closed without merging**